### ISSUE TITLE
Review logging code

### ISSUE THREAD

**Issue by DanRStevens** [OutpostUniverse/op2ext#170]
At some point the logging code should be reviewed, and perhaps made to be more regular.

Issue #155 and PR #162 made some changes to collect public logging and reporting functions into a common file. Included was some discussion of future direction.

It was planned to have global methods for each logging level:
```
void Log(...);
void LogError(...);
void LogDebug(...);
```

The current implementation looks more like:
```
void Log(const std::string& message, const std::string& moduleName = "op2ext.dll");
void PostErrorMessage(const std::string& errorMessage, const std::string& sourcePathFilename, long lineInSourceCode);
#define PostError(errorMessage) PostErrorMessage(errorMessage, __FILE__, __LINE__)
void LogDebug(const std::string& message);
```

Currently `LogError` is roughly implemented as the macro `PostError`, with the details in `PostErrorMessage`. We should try to make these consistently named. It would also be nice if we could somehow get rid of the macro irregularity. In particular, the macro requires a bit of naming irregularity since the macro name must be distinct from the underlying function name that it delegates to.

The parameters to these methods are currently different. Ideally these should be made more regular.

The `LogError` function takes a `__FILE__` and `__LINE__` location, while the others do not. By defaulting that with the macro, the remaining parameters are left more consistent. This does result in the macro irregularity however. We could also add `__FILE__` and `__LINE__` to the other methods to make them more consistent.

Additionally, the core `Log` method takes a `moduleName` parameter, which is absent from the other ones. We could perhaps push the `moduleName` detail and associated formatting up into the op2ext.cpp exported method. Direct calls to logging from within op2ext would then have a shorter parameter list. This would also free internal calls from the associated `moduleName` string formatting, which may or may not be a good thing.

The default `moduleName` parameter is set to `op2ext.dll`. This makes the code less generic and less re-usable in other projects. Moving that detail up into the op2ext.dll exported module interface could leave us with more generic and re-usable logging code.

C++20 will introduce [`source_location`](https://en.cppreference.com/w/cpp/utility/source_location). That could eventually be used to replace the `__FILE__` and `__LINE__` macro uses. The new functionality allows for much easier defaulting of the source location parameter, without having to use a macro wrapper.

We should create a logger distributor class, which can handle distributing log messages to multiple output logger destinations. This would enable a single object to output to both on screen sources, and to a log file, without having to tie the two functionalities too closely together.

The `OutputDebugString` code should be extracted into a `Logger` derived class. This would increase flexibility for debug logging.

We should add the ability to set the logger for each logging level. Currently we are missing this for the debug logging level:
```
void SetLogger(Logger* newLogger);
void SetLoggerError(Logger* newLogger);
void SetLoggerDebug(Logger* newLogger);
```

In DllMain.cpp there are multiple direct calls to the Win32 API method `OutputDebugString`. We should redirect these to `LogDebug`.

Related, though this might really be a separate issue:
The TestModule also makes direct use of the Win32 API method `OutputDebugString`.
We should review code for appropriate amounts of logging output. In particular, module load/run/unload events.
We should reduce/eliminate use of modal pop-up message boxes. (Non-modal would be much better).

**Comment by DanRStevens** [comment]
Tagging PR #235 which added the log distributor class mentioned in the first post.

----

Tagging issue #231 as related. It suggests moving the logging methods into a main `Log`/`Logger` class as static methods.

----

Tagging PR #217 for removing all references to `OutputDebugString` from TestModule.

----

I'm again thinking we can add functions to handle certain regular formatting details. For instance, including the UTC time stamp and module name for log file messages. Similarly we could have a function that adds `__FILE__` and `__LINE__` details (along with stripping the path for `__FILE__`).

Having helper formatting messages means we can deal with differences up front, before reaching the core Log API. That should allow for a simple and consistent `Log(const std::string& message)` type interface for all logging levels.

There are however some differences in the level of detail and formatting between various log destinations. For instance, a log file can persist between runs, meaning messages in it can be quite old. As such, a date/time stamp becomes much more important there. That is a detail that could be part of the log file destination itself. The current time is also easily obtained from within the log file class. Meanwhile, a message box logger would be quite current, so inclusion of date/time stamps there would be mostly just spam.

Other details such as the source module would probably be formatted into the message before logging it. Without doing that, we would need extra parameters. Further, the default module name parameter of "op2ext.dll" is rather project specific in what is otherwise rather generic code. We should handle that detail before reaching the core logging API.

I don't care to wait around for C++20 to be released and become mainstream before making improvements. We'll have to do without `source_location` for now. Such details can be handled by formatting methods. Or potentially we can just drop the file and line detail. I find I mostly just rely on easily searchable messages, rather than looking at the file and line number.

----

Searching the source code reveals the vast majority of logging is done through `PostErrorMessage`. If it's not at that level of severity, we mostly just don't log stuff.

Aside from the exported `Log` method, there is only a single internal call to `Log` from within `VolList`. Similarly, the only use of `LogDebug` is also in `VolList`.

----

Proposal:
- Add format function for `moduleName` and `message`
- Add format function for `__FILE__`, `__LINE__`, `moduleName`, and `message`
- Update `PostError` macro to first format, then log (rename `PostErrorMessage` to `LogError`)
- Use format method from exported `Log` API method, and `VolList`
- Drop the `moduleName` parameter from the `Logger` interface
- Add `LoggerDebug` class
- Add `SetLoggerDebug` function and update corresponding `LogDebug` method

That should unblock Issues #64 and #231.

**Comment by Brett208** [comment]
I'm happy if we stop supporting the line number and file the error came from. Simply using a grep search on the error message is sufficient if it isn't obvious which file the error was emitted from.

**Comment by DanRStevens** [comment]
I'm looking at potentially removing `PostError` and replacing all uses with direct calls to `LogError`. I think I might want to push that off to a new issue though, along with the impact that might have on formatting methods.
